File: src/main/java/jp/kentan/minecraft/neko_core/manager/SpawnManager.java
package jp.kentan.minecraft.neko_core.manager;

import jp.kentan.minecraft.neko_core.NekoCore;
import jp.kentan.minecraft.neko_core.component.SpawnLocation;
import jp.kentan.minecraft.neko_core.event.ConfigUpdateEvent;
import jp.kentan.minecraft.neko_core.config.SpawnConfigProvider;
import jp.kentan.minecraft.neko_core.event.SpawnCancelEvent;
import jp.kentan.minecraft.neko_core.util.Log;
import org.bukkit.*;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitScheduler;

import java.util.*;

public class SpawnManager implements SpawnCancelEvent, ConfigUpdateEvent<List<SpawnLocation>> {

    private final static BukkitScheduler SCHEDULER = Bukkit.getScheduler();

    private final Plugin PLUGIN;
    private final SpawnConfigProvider SPAWN_CONFIG;

    private Map<String, Location> SPAWN_MAP = new HashMap<>();
    private Map<Player, Integer> SPAWN_TASK_MAP = new HashMap<>();
    private List<String> SPAWN_NAME_LIST = new ArrayList<>();

    public SpawnManager(Plugin plugin, SpawnConfigProvider spawnConfigProvider) {
        PLUGIN = plugin;
        SPAWN_CONFIG = spawnConfigProvider;
    }

    public List<String> getSpawnNameList() {
        return SPAWN_NAME_LIST;
    }

    public boolean spawn(Player player, String spawnName) {
        if (!SPAWN_MAP.containsKey(spawnName)) {
            return false;
        }

        player.teleport(SPAWN_MAP.get(spawnName));
        return true;
    }

    public void addSpawnTask(Player player, String spawnName) {
        if (SPAWN_MAP.size() <= 0 || !SPAWN_MAP.containsKey(spawnName)) {
            player.sendMessage(NekoCore.PREFIX + ChatColor.YELLOW + "そのようなスポーンは存在しません.");
            return;
        }

        final Location location = SPAWN_MAP.get(spawnName);
        Location playerLocation = player.getLocation();

        player.sendMessage(NekoCore.PREFIX + "3秒後にスポーンします.");
        player.getWorld().playEffect(playerLocation, Effect.MOBSPAWNER_FLAMES, 6);
        player.playSound(playerLocation, Sound.ENTITY_PLAYER_LEVELUP, 1.0f, 0.0f);

        if (SPAWN_TASK_MAP.containsKey(player)) {
            SCHEDULER.cancelTask(SPAWN_TASK_MAP.remove(player));
        }

        int taskId = SCHEDULER.scheduleSyncDelayedTask(PLUGIN, () -> {
            if (!player.isOnline()) {
                return;
            }

            SPAWN_TASK_MAP.remove(player);
            player.teleport(location);
        }, 60L);

        SPAWN_TASK_MAP.put(player, taskId);
    }

    public void saveSpawn(Player player, String spawnName) {
        final Location location = player.getLocation();

        if(!SPAWN_CONFIG.save(spawnName, location)){
            player.sendMessage(NekoCore.PREFIX + ChatColor.RED + "保存に失敗しました.");
            return;
        }

        SPAWN_MAP.put(spawnName, location);

        if (!spawnName.equals("default")) {
            SPAWN_NAME_LIST.add(spawnName);
            Collections.sort(SPAWN_NAME_LIST);
        }

        player.sendMessage(NekoCore.PREFIX + "スポーン(" + spawnName + ")を、ここにセットしました.");
    }

    @Override
    public void onSpawnCancel(Player player) {
        if(SPAWN_TASK_MAP.containsKey(player)) {
            SCHEDULER.cancelTask(SPAWN_TASK_MAP.remove(player));
            player.sendMessage(NekoCore.PREFIX + "スポーンをキャンセルしました.");
        }
    }

    @Override
    public void onConfigUpdate(List<SpawnLocation> data) {
        SPAWN_MAP.clear();
        SPAWN_NAME_LIST.clear();

        data.forEach(spawn -> {
            SPAWN_MAP.put(spawn.NAME, spawn.LOCATION);

            if(!spawn.NAME.equals("default")) {
                SPAWN_NAME_LIST.add(spawn.NAME);
            }
        });

        Collections.sort(SPAWN_NAME_LIST);

        Log.info("SpawnMap updated.");
    }
}

File: src/main/kotlin/jp/kentan/minecraft/nekocore/manager/SpawnManager.kt
package jp.kentan.minecraft.nekocore.manager

import jp.kentan.minecraft.nekocore.NekoCorePlugin
import jp.kentan.minecraft.nekocore.data.SpawnRepository
import jp.kentan.minecraft.nekocore.util.sendSuccessMessage
import jp.kentan.minecraft.nekocore.util.sendWarnMessage
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.bukkit.Bukkit
import org.bukkit.Effect
import org.bukkit.Location
import org.bukkit.Sound
import org.bukkit.entity.Player
import java.util.concurrent.ConcurrentHashMap

class SpawnManager(
    private val plugin: NekoCorePlugin
) {

    companion object {
        const val DEFAULT_SPAWN_NAME = "default"
        const val TUTORIAL_SPAWN_NAME = "tutorial"
    }

    private val spawnRepo = SpawnRepository(plugin)

    private val spawnMap: MutableMap<String, Location> = ConcurrentHashMap()
    private val spawnTaskIdMap: MutableMap<Player, Int> = ConcurrentHashMap()

    val spawnNameList: List<String>
        get() = spawnMap.keys.toList()

    init {
        GlobalScope.launch {
            spawnMap.putAll(spawnRepo.getSpawnMap())
        }

        plugin.bukkitEventListener.subscribeSpawnCancel(::onSpawnCanceled)
        plugin.configuration.subscribe(::onConfigReloaded)
    }

    fun spawn(player: Player, name: String) {
        player.teleport(spawnMap[name] ?: return)
    }

    fun spawnDelay(player: Player, name: String?) {
        val location = spawnMap[name ?: DEFAULT_SPAWN_NAME] ?: let {
            player.sendWarnMessage("そのようなスポーンは存在しません.")
            return
        }

        player.apply {
            sendMessage("${NekoCorePlugin.PREFIX}3秒後にスポーンします.")
            world.playEffect(this.location, Effect.MOBSPAWNER_FLAMES, 6)
            playSound(this.location, Sound.ENTITY_PLAYER_LEVELUP, 1.0f, 0.0f)
        }

        val prevTaskId = spawnTaskIdMap[player]
        if (prevTaskId != null) {
            Bukkit.getScheduler().cancelTask(prevTaskId)
        }

        val taskId = Bukkit.getScheduler().scheduleSyncDelayedTask(plugin, {
            spawnTaskIdMap.remove(player)

            if (player.isOnline) {
                player.teleport(location)
            }
        }, 60L)

        spawnTaskIdMap[player] = taskId
    }

    fun setSpawn(player: Player, name: String) {
        GlobalScope.launch {
            val location = player.location

            if (spawnRepo.addSpawn(name, location)) {
                spawnMap[name] = location

                player.sendSuccessMessage("スポーン($name)をここに設定しました.")
            } else {
                player.sendWarnMessage("スポーンの保存に失敗しました.")
            }
        }
    }

    private fun onSpawnCanceled(player: Player) {
        val taskId = spawnTaskIdMap.remove(player) ?: return

        Bukkit.getScheduler().cancelTask(taskId)
        player.sendMessage("${NekoCorePlugin.PREFIX}スポーンをキャンセルしました.")
    }

    private fun onConfigReloaded() {
        GlobalScope.launch {
            spawnMap.apply {
                val newSpawnMap = spawnRepo.getSpawnMap()
                clear()
                putAll(newSpawnMap)
            }
        }
    }
}